File: cli/src/fp_cli/commands/deploy.py
"""fp deploy — deploy to the platform."""

import time

import typer
from rich.live import Live
from rich.spinner import Spinner

from ..api.client import PlatformClient, PlatformError, resolve_app
from ..config import get_active_platform_or_exit
from ..console import console, err_console
from ..project import read_project, collect_files, detect_mode

PHASE_LABELS = {
    "validating": "Validating code...",
    "creating_resources": "Creating Kubernetes resources...",
    "pending": "Waiting to be scheduled...",
    "scheduled": "Pod scheduled...",
    "pulling": "Pulling container image...",
    "pulled": "Image pulled...",
    "creating": "Creating container...",
    "starting": "Starting application...",
    "ready": "Application ready!",
    "error": "Deployment error",
}


def deploy():
    """Deploy the current project to the platform."""
    project = read_project()
    app_name = project.get("name")
    entrypoint = project.get("entrypoint", "app.py")

    if not app_name:
        err_console.print("[red]No 'name' field in .fp.yaml[/red]")
        raise typer.Exit(1)

    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    # Collect files from directory
    files = collect_files(entrypoint=entrypoint)
    mode = detect_mode(files)

    console.print(f"Deploying [bold]{app_name}[/bold] to {platform['url']}")
    if mode == "multi":
        console.print(f"  {len(files)} files, mode: multi-file")
    else:
        console.print(f"  mode: single-file")

    # Check if app already exists (update vs create)
    existing_apps = client.list_apps()
    existing = next((a for a in existing_apps if a["name"] == app_name), None)

    # Build payload
    payload: dict = {"name": app_name}
    if mode == "multi":
        payload["files"] = files
        payload["mode"] = "multi"
        payload["entrypoint"] = entrypoint
    else:
        payload["code"] = files[entrypoint]
        payload["mode"] = "single"

    # Include env vars from .fp.yaml if present
    env = project.get("env")
    if env and isinstance(env, dict):
        import os

        resolved = {}
        for key, value in env.items():
            if isinstance(value, str) and value.startswith("${") and value.endswith("}"):
                env_name = value[2:-1]
                resolved[key] = os.environ.get(env_name, "")
            else:
                resolved[key] = str(value)
        payload["env_vars"] = resolved

    try:
        if existing:
            result = client.update_app(existing["app_id"], payload)
            console.print(f"  Updating existing app ({existing['app_id']})")
        else:
            result = client.create_app(payload)
            console.print(f"  Created new app ({result.get('app_id', 'unknown')})")
    except PlatformError as e:
        err_console.print(f"\n[red]Deploy failed:[/red] {e.message}")
        raise typer.Exit(1)

    app_id = result.get("app_id") or (existing and existing["app_id"])
    if not app_id:
        err_console.print("[red]No app_id in response[/red]")
        raise typer.Exit(1)

    # Poll deployment status
    _poll_deploy_status(client, app_id, platform["url"])


def _poll_deploy_status(client: PlatformClient, app_id: str, platform_url: str):
    """Poll deploy-status until ready, error, or timeout."""
    max_polls = 60
    poll_interval = 2

    with Live(
        Spinner("dots", text="Creating resources..."),
        console=console,
        transient=True,
    ) as live:
        for i in range(max_polls):
            try:
                status = client.deploy_status(app_id)
            except PlatformError:
                time.sleep(poll_interval)
                continue

            phase = "creating_resources"
            if status.get("deployment_ready"):
                phase = "ready"
            elif status.get("status") == "error":
                phase = "error"
            elif status.get("deploy_stage"):
                phase = status["deploy_stage"]

            label = PHASE_LABELS.get(phase, phase)
            live.update(Spinner("dots", text=label))

            if phase == "ready":
                break
            if phase == "error":
                err_console.print(
                    f"\n[red]Deployment failed:[/red] {status.get('last_error', 'Unknown error')}"
                )
                raise typer.Exit(1)

            time.sleep(poll_interval)
        else:
            err_console.print(
                "\n[yellow]Deployment is still in progress.[/yellow] "
                "Check status with [bold]fp status[/bold]"
            )
            raise typer.Exit(1)

    # Derive app URL from platform URL
    # platform.gatorlunch.com -> app-{id}.gatorlunch.com
    from urllib.parse import urlparse

    parsed = urlparse(platform_url)
    host = parsed.hostname or ""
    # Strip "platform." prefix to get the app domain
    if host.startswith("platform."):
        app_domain = host[len("platform."):]
    else:
        app_domain = host
    app_url = f"https://app-{app_id}.{app_domain}"

    console.print()
    console.print("[green]Deployed successfully![/green]")
    console.print(f"  URL: [bold link={app_url}]{app_url}[/bold link]")
    console.print(f"  Docs: [link={app_url}/docs]{app_url}/docs[/link]")

File: cli-go/cmd/deploy.go
package cmd

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/api"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/config"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/project"
)

var deployCmd = &cobra.Command{
	Use:   "deploy",
	Short: "Deploy the current project to the platform",
	RunE:  runDeploy,
}

var phaseLabels = map[string]string{
	"validating":         "Validating code...",
	"creating_resources": "Creating Kubernetes resources...",
	"pending":            "Waiting to be scheduled...",
	"scheduled":          "Pod scheduled...",
	"pulling":            "Pulling container image...",
	"pulled":             "Image pulled...",
	"creating":           "Creating container...",
	"starting":           "Starting application...",
	"ready":              "Application ready!",
	"error":              "Deployment error",
}

func runDeploy(cmd *cobra.Command, args []string) error {
	proj, dir, err := project.ReadFromCwd()
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	if proj.Name == "" {
		fmt.Fprintln(os.Stderr, "No 'name' field in .fp.yaml")
		os.Exit(1)
	}

	p := config.GetActivePlatform()
	if p == nil {
		fmt.Fprintln(os.Stderr, "Not authenticated. Run 'fp auth login <platform-url>' first.")
		os.Exit(1)
	}

	entrypoint := proj.Entrypoint
	if entrypoint == "" {
		entrypoint = "app.py"
	}

	files, err := project.CollectFiles(dir, entrypoint)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	mode := project.DetectMode(files)
	client := api.NewClient(p.URL, p.Token)

	fmt.Printf("Deploying %s to %s\n", proj.Name, p.URL)
	if mode == "multi" {
		fmt.Printf("  %d files, mode: multi-file\n", len(files))
	} else {
		fmt.Println("  mode: single-file")
	}

	apps, err := client.ListApps()
	if err != nil {
		return err
	}
	var existing map[string]interface{}
	for _, a := range apps {
		if n, _ := a["name"].(string); n == proj.Name {
			existing = a
			break
		}
	}

	payload := make(map[string]interface{})
	payload["name"] = proj.Name
	if mode == "multi" {
		payload["files"] = files
		payload["mode"] = "multi"
		payload["entrypoint"] = entrypoint
	} else {
		payload["code"] = files[entrypoint]
		payload["mode"] = "single"
	}

	if proj.Env != nil {
		resolved := make(map[string]string)
		for k, v := range proj.Env {
			if strings.HasPrefix(v, "${") && strings.HasSuffix(v, "}") {
				resolved[k] = os.Getenv(v[2 : len(v)-1])
			} else {
				resolved[k] = v
			}
		}
		payload["env_vars"] = resolved
	}

	var result map[string]interface{}
	if existing != nil {
		appID, _ := existing["app_id"].(string)
		result, err = client.UpdateApp(appID, payload)
		if err != nil {
			return handleAPIError(err)
		}
		fmt.Printf("  Updating existing app (%s)\n", appID)
	} else {
		result, err = client.CreateApp(payload)
		if err != nil {
			return handleAPIError(err)
		}
		appID, _ := result["app_id"].(string)
		fmt.Printf("  Created new app (%s)\n", appID)
	}

	appID, _ := result["app_id"].(string)
	if appID == "" && existing != nil {
		appID, _ = existing["app_id"].(string)
	}
	if appID == "" {
		fmt.Fprintln(os.Stderr, "No app_id in response")
		os.Exit(1)
	}

	// Poll deploy status
	for i := 0; i < 60; i++ {
		status, err := client.DeployStatus(appID)
		if err != nil {
			time.Sleep(2 * time.Second)
			continue
		}

		phase := "creating_resources"
		if ready, _ := status["deployment_ready"].(bool); ready {
			phase = "ready"
		} else if s, _ := status["status"].(string); s == "error" {
			phase = "error"
		} else if stage, ok := status["deploy_stage"].(string); ok {
			phase = stage
		}

		label := phaseLabels[phase]
		if label == "" {
			label = phase
		}
		fmt.Printf("\r%s", label)

		if phase == "ready" {
			break
		}
		if phase == "error" {
			lastErr, _ := status["last_error"].(string)
			if lastErr == "" {
				lastErr = "Unknown error"
			}
			fmt.Fprintf(os.Stderr, "\nDeployment failed: %s\n", lastErr)
			os.Exit(1)
		}

		time.Sleep(2 * time.Second)
	}

	appURL := api.AppURL(p.URL, appID)
	fmt.Println()
	fmt.Println("Deployed successfully!")
	fmt.Printf("  URL:  %s\n", appURL)
	fmt.Printf("  Docs: %s/docs\n", appURL)
	return nil
}

func handleAPIError(err error) error {
	var pe *api.PlatformError
	if errors.As(err, &pe) {
		fmt.Fprintf(os.Stderr, "\nDeploy failed: %s\n", pe.Message)
		os.Exit(1)
	}
	return err
}
